### DISCUSSION TITLE
Response/Error Formatting for cURL/Postman

### DISCUSSION THREAD

**Discussion by PointSingularity** [trpc/trpc#2466] [Q&A] [answered]
Thanks for the great library!

I have a tRPC(v9) application working with Express/React. 
When a request with bad data is sent, the error is shown nicely in the console, or through accessing the error object.
```jsx
<pre>{mutation.error.message}</pre>
```
![image](https://user-images.githubusercontent.com/56246797/184842793-49b19c00-7484-41cd-bd69-e07cec30c79f.png)
![image](https://user-images.githubusercontent.com/56246797/184842954-4d78c5fe-89a9-463a-b1dc-66975c5a4aa8.png)

Sometimes I use cURL/Invoke-RestMethod/Postman to test some endpoints. The errors there have all the escape characters present.

![image](https://user-images.githubusercontent.com/56246797/184843080-e9792d16-8d34-4f9a-9fd0-1f411de3f540.png)
![image](https://user-images.githubusercontent.com/56246797/184843258-1274c1f5-c162-4c69-beec-9218533837d8.png)

Can this formatting be disabled?

Any insight would be helpful, thank you in advance!

**Comment by sachinraja** [comment, marked as answer]
Whatever error you're getting is from the validation library you're using, so you should look there to configure the formatting.

**Reply by PointSingularity** [reply]
Thank you for your answer, I've solved the formatting issues using "formatError".

Follow-up question:
I have multiple routers defined, and I want to do some specific errors based on the route.
If add this to a specific route, it doesn't work:
```typescript
  .formatError(({ shape, error }) => {
    return {
      message:
        shape.data.code === 'BAD_REQUEST'
          ? `The request wasn't properly shaped. Check the data property for a proper request object.`
          : shape.message,
      code: shape.code,
      data: {
        properFormatting: sidObject,
        specificErrors:
          error.code === 'BAD_REQUEST' && error.cause instanceof ZodError
            ? error.cause.flatten()
            : null,
      },
    };
  });
```
But if add it to the main router where all of them are merged, it works:
```typescript
  .formatError(({ shape, error }) => {
    return {
      message:
        shape.data.code === 'BAD_REQUEST' && shape.data.path === 'sid'
          ? `The request wasn't properly shaped. Check the data property for a proper request object.`
          : shape.message,
      code: shape.code,
      data: {
        properFormatting: {},
        specificErrors:
          error.code === 'BAD_REQUEST' && error.cause instanceof ZodError
            ? error.cause.flatten()
            : null,
      },
    };
  });
```
If I did it for every route, or edge case, it would get really long.
Not sure if I am doing something wrong.

**Reply by PointSingularity** [reply]
@sachinraja sorry to be a bother, but could you provide some input on the issue above? Or should I maybe open a new discussion?

**Reply by sachinraja** [reply]
This is because you can only have one error formatter in your router tree (I think due to batching). Child routers will be overridden by their parent routers.

**Reply by PointSingularity** [reply]
> This is because you can only have one error formatter in your router tree (I think due to batching). Child routers will be overridden by their parent routers.

Gotcha, thanks for the answer. 
So there isn't any other way to achieve custom error formats per route (I haven't checked v10 yet)?

**Reply by sachinraja** [reply]
No I don't think so unfortunately. On v10, `formatError` was moved to a top level [`errorFormatter`](https://alpha.trpc.io/docs/error-formatting) so you can only declare one.

**Reply by PointSingularity** [reply]
Thanks for the info!
The top level errorFormatter would then make it easier to define it in a separate file, which makes it more manageable. 
I'll probably migrate to it then 👍
